sdk/identity/identity/src/credentials/deviceCodeCredential.ts

```diff
@@ -0,0 +1,203 @@
+    deviceCodeResponse: DeviceCodeResponse,
+    options?: GetTokenOptions
+  ): Promise<TokenResponse | null> {
+    let tokenResponse: TokenResponse | null = null;
+
+    const webResource = this.identityClient.createWebResource({
+      url: `${this.identityClient.authorityHost}/${this.tenantId}/oauth2/v2.0/token`,
+      method: "POST",
+      disableJsonStringifyOnBody: true,
+      deserializationMapper: undefined,
+      body: qs.stringify({
+        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
+        client_id: this.clientId,
+        device_code: deviceCodeResponse.device_code
+      }),
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/x-www-form-urlencoded"
+      },
+      abortSignal: options && options.abortSignal
+    });
+
+    while (tokenResponse === null) {
+      try {
+        await delay(deviceCodeResponse.interval * 1000);
+
+        // Check the abort signal before sending the request
+        if (options && options.abortSignal && options.abortSignal.aborted) {
+          return null;
+        }
+
+        tokenResponse = await this.identityClient.sendTokenRequest(webResource);
+      } catch (err) {
+        if (err instanceof AuthenticationError) {
+          switch (err.errorResponse.error) {
+            case "authorization_pending":
+              break;
+            case "authorization_declined":
+              return null;
+            case "expired_token":
+              throw err;
+            case "bad_verification_code":
+              throw err;
+          }
+        } else {
+          throw err;
+        }
+      }
+    }
+
+    return tokenResponse;
+  }
+
+  /**
+   * Authenticates with Azure Active Directory and returns an {@link AccessToken} if
+   * successful.  If authentication cannot be performed at this time, this method may
+   * return null.  If an error occurs during authentication, an {@link AuthenticationError}
+   * containing failure details will be thrown.
+   *
+   * @param scopes The list of scopes for which the token will have access.
+   * @param options The options used to configure any requests this
+   *                TokenCredential implementation might make.
+   */
+  public async getToken(
+    scopes: string | string[],
+    options?: GetTokenOptions
+  ): Promise<AccessToken | null> {
+    let tokenResponse: TokenResponse | null = null;
+    let scopeString = typeof scopes === "string" ? scopes : scopes.join(" ");
+    if (scopeString.indexOf("offline_access") < 0) {
+      scopeString += " offline_access";
+    }
+
+    // Try to use the refresh token first
+    if (this.lastTokenResponse && this.lastTokenResponse.refreshToken) {
+      tokenResponse = await this.identityClient.refreshAccessToken(
+        this.tenantId,
+        this.clientId,
+        scopeString,
+        this.lastTokenResponse.refreshToken,
+        undefined, // clientSecret not needed for device code auth
+        undefined,
+        options);
```

Comment:
it's possible that this could error.  In fact when the refresh token is expired it will pass back an error code.  This needs to be handled here or in the identityClient.

Reply:
Yep, I'll add checks for the refresh-specific errors